README.md
<img align="right" width="150" alt="Logo" src="docs/assets/era5cli_logo_colors_border.png">


[![github license badge](https://img.shields.io/github/license/eWaterCycle/era5cli)](https://github.com/eWaterCycle/era5cli)
[![rsd badge](https://img.shields.io/badge/RSD-era5cli-blue)](https://research-software-directory.org/software/era5cli)
[![fair-software.eu](https://img.shields.io/badge/fair--software.eu-%E2%97%8F%20%20%E2%97%8F%20%20%E2%97%8F%20%20%E2%97%8F%20%20%E2%97%8B-yellow)](https://fair-software.eu)
[![DOI](https://zenodo.org/badge/DOI/10.5281/zenodo.3252665.svg)](https://doi.org/10.5281/zenodo.3252665)

[![Documentation Status](https://readthedocs.org/projects/era5cli/badge/?version=stable)](https://lilio.readthedocs.io/en/stable/?badge=stable)
[![build](https://github.com/eWaterCycle/era5cli/actions/workflows/test_and_build.yml/badge.svg)](https://github.com/eWaterCycle/era5cli/actions/workflows/test_and_build.yml)
[![Test Coverage](https://codecov.io/gh/eWaterCycle/era5cli/branch/main/graph/badge.svg?token=qeZXgGASBK)](https://codecov.io/gh/eWaterCycle/era5cli)
[![PyPI](https://badge.fury.io/py/era5cli.svg)](https://badge.fury.io/py/era5cli)


A command line interface to download ERA5 data from the [Copernicus Climate Data Store](<https://climate.copernicus.eu/>).

<hr>

With `era5cli` you can:

 - Download meteorological data in GRIB/NetCDF, including ERA5 data from the preliminary back extension, and ERA5-Land data.
 - List and retrieve information on available variables and pressure levels
 - Select multiple variables for several months and years
 - Split outputs by years (and optionally months), producing a separate files instead of merging them in one file
 - Download multiple files at once
 - Extract data for a sub-region of the globe

<hr>

Free software: Apache Software License 2.0

Documentation: https://era5cli.readthedocs.io



CHANGELOG.rst
Release Notes
*************

Unreleased
~~~~~~~~~~
* Change CDS keys from .cdsapirc file to .config/eracli.txt file. This will avoid conflict with e.g. ADS.
* Add validator for era5cli.txt keys. This should provide better feedback to users and reduce user error.
* Added --splitmonths argument for `era5cli hourly`. This allows users to avoid a Request Too Large error.
* If a user makes a request without --splitmonths they are warned that the behavior will change in the future, and that they have to choose between --splitmonths False and --splitmonths True.
* When a request would encounter a Request Too Large error in the CDS API, they are warned, and given a suggestion to use --splitmonths.
* cli.py has been refactored to make the structure more clear. Seperate argument builders are now in their own modules.

1.3.2 (2021-12-13)
~~~~~~~~~~~~~~~~~~
* Elaborate the range of years that can be queried `#123 <https://github.com/eWaterCycle/era5cli/pull/123>`_
* Update Readthedocs theme `#125 <https://github.com/eWaterCycle/era5cli/pull/125>`_
* Fix a bug that allowed the incompatible combination of --land and --ensemble `#131 <https://github.com/eWaterCycle/era5cli/pull/131>`_

1.3.1 (2021-12-01)
~~~~~~~~~~~~~~~~~~
* Automatic Zenodo/RSD release failed; updated contribution guidelines `#106 <https://github.com/eWaterCycle/era5cli/pull/106>`_

1.3.0 (2021-11-30)
~~~~~~~~~~~~~~~~~~
* Fix compatibility with changed CDS variables geopotential/orography `#98 <https://github.com/eWaterCycle/era5cli/pull/98>`_
* Add integration testing `#102 <https://github.com/eWaterCycle/era5cli/pull/102>`_

1.2.1 (2021-04-21)
~~~~~~~~~~~~~~~~~~
* Automatic PyPI release for 1.2.0 failed; updated github action workflow `#91 <https://github.com/eWaterCycle/era5cli/pull/91>`_

1.2.0 (2021-04-21)
~~~~~~~~~~~~~~~~~~
* Add support for ERA5-Land data `#67 <https://github.com/eWaterCycle/era5cli/pull/67>`_
* Add functionality to download subregions `#70 <https://github.com/eWaterCycle/era5cli/pull/70>`_
* Update variables available for ERA5 datasets `#84 <https://github.com/eWaterCycle/era5cli/pull/84>`_

1.1.1 (2020-12-15)
~~~~~~~~~~~~~~~~~~
* Patch to fix the github actions publish automation `#64 <https://github.com/eWaterCycle/era5cli/pull/64>`_

1.1.0 (2020-12-14)
~~~~~~~~~~~~~~~~~~
* The stable 1.1.0 era5cli minor release.
* Add support for ERA5 preliminary back extension `#58 <https://github.com/eWaterCycle/era5cli/pull/58>`_
* Update tests `#57 <https://github.com/eWaterCycle/era5cli/pull/57>`_
* Add automated PyPI package building and publishing with github Actions `#62 <https://github.com/eWaterCycle/era5cli/pull/62>`_

1.0.0 (2019-07-25)
~~~~~~~~~~~~~~~~~~
* The stable 1.0.0 era5cli release.
* Adding more useful information to netCDF history `#48 <https://github.com/eWaterCycle/era5cli/pull/48>`_

1.0.0rc3 (2019-07-16)
~~~~~~~~~~~~~~~~~~~~~
* Third Release Candidate for the stable 1.0.0 era5cli release.
* Improve documentation `#21 <https://github.com/eWaterCycle/era5cli/issues/21>`_ `#29 <https://github.com/eWaterCycle/era5cli/issues/29>`_.
* Cleanup command line options `#19 <https://github.com/eWaterCycle/era5cli/issues/19>`_ `#20 <https://github.com/eWaterCycle/era5cli/issues/20>`_.
* Append era5cli version to history of downloaded netCDF file `#17 <https://github.com/eWaterCycle/era5cli/issues/17>`_.

1.0.0rc2 (2019-07-01)
~~~~~~~~~~~~~~~~~~~~~
* Second Release Candidate for the stable 1.0.0 era5cli release.
* Fix downloading all variables when requesting multiple variables and using --split `#23 <https://github.com/eWaterCycle/era5cli/issues/23>`_.
* Fix link to PyPI package in documentation `#22 <https://github.com/eWaterCycle/era5cli/issues/22>`_.

1.0.0rc1 (2019-06-22)
~~~~~~~~~~~~~~~~~~~~~
* First Release Candidate for the stable 1.0.0 era5cli release: A commandline utility to download ERA-5 data using cdsapi.


pyera5_download.py
'''
description:    Download ERA5 from 
license:        APACHE 2.0
author:         Yifat Dzigan, NLeSC (y.dzigan@esciencecenter.nl)
'''

from optparse import OptionParser
import sys
import cdsapi
import re


usage = """Usage:

    --years <years> --variable <variable> 

    Examples:

    runoff for the year 2018
    >>> python3 --years 2018 --variable runoff

    runoff for 2017 and 2018
    >>> python3 --years 2017,2018 --variable runoff

    
""".format(__file__)
parser = OptionParser(usage = usage)
parser.add_option("-y", "--years", dest = "years", type = "str",
        help="years for which the data should be downloaded. " + \
                 "Can be a single year or a comma separated list ")
parser.add_option("-p", "--variable", dest = "var", type = "str",
       help="the variable to be downloaded. See the cds website for availabe variables.")
(options,args) = parser.parse_args()


if options.years == None or options.var == None:
    parser.print_help(); sys.exit(9)
else:
    if re.match("^[0-9]{4}$", options.years):
        options.years = [int(options.years)]
    elif re.match("^[0-9,]+$", options.years):
        options.years = [int(x) for x in options.years.split(",")]
    else:
        parser.print_help(); log.info("\n\n")
        raise ValueError("Wrong format for -y/--years")

HydroVariable = options.var
years = options.years

c = cdsapi.Client()
c.retrieve('reanalysis-era5-single-levels', {
        	'variable'      : HydroVariable,
        'product_type'  : 'reanalysis',
        'year'          : years,
        'month'         : ['01','02','03','04','05','06',
            		   '07','08','09','10','11','12'],
        'day'           : '01',
        'time'          : [
            '00:00','01:00','02:00',
            '03:00','04:00','05:00',
            '06:00','07:00','08:00',
            '09:00','10:00','11:00',
            '12:00','13:00','14:00',
            '15:00','16:00','17:00',
            '18:00','19:00','20:00',
            '21:00','22:00','23:00'
        ],
        'format'        : 'netcdf' 
    }, 'test.nc')


